erlang: update to 25.3

hgy59 opened the pull request:
## Description

This PR updates the Erlang ecosystem packages (Erlang, Elixir, RabbitMQ, ejabberd) and fixes critical DSM 7 upgrade issues.

### Version Updates

| Package | Old Version | New Version |
|---------|-------------|-------------|
| Erlang | 25.0.4 | 25.3.2.21 |
| Elixir (native) | 1.14.0 | 1.15.8 |
| RabbitMQ | 3.10.7 | 3.11.28 |
| ejabberd | 21.07 | 23.10 |

### Key Improvements

**ejabberd**
- Fixed configuration paths for DSM 7+ (`SYNOPKG_PKGVAR` instead of `SYNOPKG_PKGDEST`)
- Added dynamic upgrade wizard that detects missing admin ACL and prompts for credentials
- Added automatic data migration from old paths during DSM 7 upgrades
- Added CLI wrapper with proper permission handling for `ejabberdctl`
- Added web interface shortcut in DSM

**RabbitMQ**
- Fixed configuration for DSM 7
- Added DSM 7 upgrade data migration (cookie and var folder)
- Added CLI wrappers with proper permission handling for all management tools
- Added web interface shortcut to RabbitMQ Management UI (port 15672)

**Erlang**
- Fixed OpenSSL include order to resolve build issues
- Disabled parallel make for Elixir bootstrap to fix race condition


## Checklist

- [x] Build rule `all-supported` completed successfully
- [x] New installation of package completed successfully
- [x] Package upgrade completed successfully (Manually install the package again)
- [x] Package [functionality was tested](https://github.com/SynoCommunity/spksrc/wiki/Package-Update-Policy#tests-checks)
- [x] Any needed [documentation](https://github.com/SynoCommunity/spksrc/wiki/Create-documentation) is updated/created


### Type of change

<!--Please use any relavent tags.-->
- [x] Package update

hgy59 committed:
erlang: update to 25.3

- update erlang to v25.3.2.5
- includes update to cross/openssl3

hgy59 committed:
rabbitmq: update to v3.11.20

- add a git tag for the time of the build as included packages evaluate the rabbitmq version with "git describe --tag"

hgy59 committed:
ejabberd: update to v23.04

hgy59 committed:
ejabberd: fix packaging of config files

hgy59 committed:
ejabberd: provide a copy of the new config files on upgrade

hgy59 committed:
Merge branch 'master' of https://github.com/hgy59/spksrc into update_erlang

mreid-tt commented:
hey @hgy59, may I add some commits to your PR to get the build to work again?

mreid-tt committed:
native/elixir: disable parallel make to fix bootstrap race condition

Elixir's Makefile has .NOTPARALLEL: compile but spksrc's make -j
overwrites this, causing race conditions during bootstrap that
result in 'init terminating in do_boot' errors in CI.

mreid-tt committed:
Update erlang to 25.3.2.7, rabbitmq to 3.11.28, ejabberd to 23.10

mreid-tt committed:
erlang: update to 25.3.2.21 with OpenSSL include fix

Adds patch to fix erl_interface and erts/emulator build when
cross-compiling. The issue is that the staging directory's OpenSSL
headers get picked up before Erlang's custom headers, preventing
proper MD5 function renaming (ei_MD5Init/ei_MD5Update/ei_MD5Final).

The patch prepends -Iopenssl/include to CFLAGS to ensure Erlang's
custom OpenSSL headers are found first.

mreid-tt committed:
native/elixir: update to 1.15.8

hgy59 commented:
> hey @hgy59, may I add some commits to your PR to get the build to work again?

Yes, please, if you have a solution...

hgy59 commented:
> * native/erlang: update to 25.3.2.21

This can stay on 25.3 as before. It must not match the cross/erlang version (this is different to other systems like native/python*).
Only major.minor version must match, patch version is not relevant.

mreid-tt commented:
> This can stay on 25.3 as before. It must not match the cross/erlang version (this is different to other systems like native/python*).
> Only major.minor version must match

Can you explain why this may be? Happy to roll it back but was curious about this.

mreid-tt committed:
native/erlang: roll-back to 25.3

hgy59 committed:
fix rabbitmq configuration for DSM 7

mreid-tt committed:
ejabberd: fix config path for DSM 7+

Fix ejabberdctl paths to use /var/packages/ejabberd/var instead of
/var/packages/ejabberd/target/var for CONFIG_DIR, LOGS_DIR, and
SPOOL_DIR.

This uses the INSTALL_PREFIX_VAR variable which is correctly set by
the framework for DSM 7+ packages.

hgy59 committed:
fix INSTALL_PREFIX_VAR

- use the same approach for all DSM versions

hgy59 committed:
ejabberd: fixes

- install several config files to var folder
- use bin instead of sbin folder to avoid additional patch
- add missing artifacts to PLIST
- PLIST: remove files in var folder
- configure web admin authentication in ejabberd.yml

hgy59 commented:
> DSM7's var environment is indeed different,

The generic installer of spksrc ensures the `$(INSTALL_DIR)/var` folder for all DSM versions.
So all DSM version have a DSM 7 compatible var folder.

th0ma7 commented on mk/spksrc.directories.mk:
## TL;DR;
@hgy59 i started writing this being absolutely sure you were breaking things up for DSM6 with this small change... further i dig into it less I'm sure its the case - a few install + upgrade testing is needed to confirm its actually all is good.

EDIT: `spksrc.copy.mk` also needs changing if you want to enable this behavior by default...

## Long-version:
DSM6 = `/var/packages/<name>/target/var`
DSM7 = `/var/packages/<name>/var`

With the following change you're enforcing DSM6 `var` to be installed in the wrong path... as the `@appdata` does not exist.

Although interestingly now DSM6 allow using `/var/packages/<name>/var` but in reality DSM6 `var` symlink actually point to `target/var` ... I don't recall this was the case when I made this change for DSM7 a while back but that might have been an oversight of me.

`git` on DSM7:
```
th0ma7@th0ma7-nas:/var/packages/git$ ls -la
total 40
drwxr-xr-x  5 root root  4096 Nov 22 12:09 .
drwxr-xr-x 72 root root  4096 Jan  4 12:09 ..
drwxr-xr-x  2 root root  4096 Nov 22 12:08 conf
-rw-r--r--  1 root root     0 Nov 22 12:09 enabled
lrwxrwxrwx  1 root root    21 Oct 22 06:14 etc -> /volume1/@appconf/git
lrwxrwxrwx  1 root root    21 Oct 22 06:14 home -> /volume1/@apphome/git
-rw-r--r--  1 root root 17417 Oct 22 06:14 INFO
drwxr-xr-x  2 root root  4096 Oct 20 15:31 scripts
lrwxrwxrwx  1 root root    22 Oct 22 06:14 share -> /volume1/@appshare/git
lrwxrwxrwx  1 root root    22 Oct 22 06:14 target -> /volume1/@appstore/git
lrwxrwxrwx  1 root root    21 Oct 22 06:14 tmp -> /volume1/@apptemp/git
lrwxrwxrwx  1 root root    21 Oct 22 06:14 var -> /volume1/@appdata/git
drwxr-xr-x  2 root root  4096 Oct 20 15:31 WIZARD_UIFILES
```
And `git` on DSM6:
```
th0ma7@th0ma7-armv5:/var/packages/git$ ls -la
total 36
drwxr-xr-x 4 root root  4096 Jan 10 21:13 .
drwxr-xr-x 6 root root  4096 Jan 10 21:13 ..
drwxr-xr-x 2 root root  4096 Jan 10 21:13 conf
-rw-r--r-- 1 root root     0 Jan 10 21:13 enabled
lrwxrwxrwx 1 root root    26 Jan 10 21:13 etc -> /usr/syno/etc/packages/git
-rw-r--r-- 1 root root 17622 Jan 10 21:13 INFO
drwxr-xr-x 2 root root  4096 Oct 20 14:31 scripts
lrwxrwxrwx 1 root root    22 Jan 10 21:13 target -> /volume1/@appstore/git
lrwxrwxrwx 1 root root    26 Jan 10 21:13 var -> /volume1/@appstore/git/var
```

Issue is, there isn't any migration operation from `target/var` to `target/../var` in the DSM6 install scripts.  I first thought this was going to break existing packages upon their upgrade... but maybe not... 

dsm7 script has a `syno_sync_var_folder` which migrates the data from `target/var` to `var` ... but as the data is already in the right location in the package (i.e. already is in `target/var`), in theory no move is needed.  And since the symlink exists, paths expected at compile time should still work, surprisingly ?!??

And also suggesting a clearer code for `$(shell dirname $(INSTALL_PREFIX))/var` 
to be using instead `$(abspath $(INSTALL_PREFIX)/../var)`

Further digging into it... spksrc.copy.mk does a hook to capture this special `var` directory such as:
```
copy_target: SHELL:=/bin/bash
copy_target: .SHELLFLAGS := -o pipefail -c
copy_target: $(PRE_COPY_TARGET) $(INSTALL_PLIST)
ifeq ($(call version_ge, ${TCVERSION}, 7.0),1)
        @$(MSG) [DSM7+] Copy target to staging, discard var directory
        @(mkdir -p $(STAGING_DIR) && cd $(STAGING_INSTALL_PREFIX) && tar cpf - $$(cat $(INSTALL_PLIST) | sed -e '/^.*:var\/.*/d' -e 's/^.*://g')) | \
          tar xpf - -C $(STAGING_DIR)
        @$(MSG) "[DSM7+] Copy and merge var and target/var to $(STAGING_DIR)/var"
        @if [ "$$(cat $(INSTALL_PLIST) | sed -n 's?^.*:var/??p')" ] ; then \
          mkdir -p $(STAGING_DIR)/var ; \
          (cd $(STAGING_INSTALL_PREFIX)/../ && tar cpf - $$(eval ls -d $$(cat $(INSTALL_PLIST) | sed -n 's?^.*:var/??p' | sed -e 's?^?{var,target/var}/?') 2>/dev/null)) | \
            tar xpf - -C $(STAGING_DIR)/var --strip-components=1 --transform='s!^target/!!' ; \
        fi
else
        @$(MSG) [DSM6] Copy target to staging
        @(mkdir -p $(STAGING_DIR) && cd $(STAGING_INSTALL_PREFIX) && tar cpf - $$(cat $(INSTALL_PLIST) | cut -d':' -f2)) | tar xpf - -C $(STAGING_DIR)
endif
```
This means that under DSM6 these files would not be packaged... thus `copy.mk` needs changing as well.

th0ma7 replied:
> we have the links for DSM 5 and 6 since #4727 And I can't remember any issue. The packages must ensure to use /var/packages/packagename/var and not the resolved link in the code an in scripts anyway.

Gotcha, I figure they didn't existed when I added the var facility management.

> IMHO copy_target should not need to handle a target/var folder, no package should have files from cross Makefile in the target/var folder (i.e. having :var in PLIST), such files must be copied by the spk Makefile (extra_install) directly to the staging dir.

I disagree, while rather rare there are files provided from some of the cross such as:
```
spksrc@spksrc13:~/master/spksrc/cross$ grep var */PLIST
ejabberd/PLIST:rsc:var/ejabberd.yml
ejabberd/PLIST:rsc:var/ejabberdctl.cfg
ejabberd/PLIST:rsc:var/inetrc
imapfilter/PLIST:rsc:var/config.lua
imapfilter/PLIST:rsc:var/extend.lua
intel-libvpl/PLIST:rsc:etc/vpl/vars.sh
lm-sensors/PLIST:rsc:var/sensors3.conf
mutt/PLIST:rsc:var/Muttrc
mutt/PLIST:rsc:var/mime.types
mutt/PLIST:rsc:var/mime.types.dist
phpmemcachedadmin/PLIST:rsc:var/phpmemcachedadmin.config/
rsnapshot/PLIST:rsc:var/rsnapshot.conf.default
saltgui/PLIST:rsc:var/saltgui/**
slang/PLIST:lib:lib/slang/v2/modules/varray-module.so
smartmontools/PLIST:rsc:var/smartd.conf
smartmontools/PLIST:rsc:var/smartd_warning.sh
zap2epg/PLIST:rsc:var/epggrab/zap2epg/conf/zap2epg.xml
```

> The only difference is that under DSM 6 providing new file under var as *.new on package upgrade must be done explicit (like in service_preinst of ejabberd).

That's exactly what's being done automatically in `spksrc.service.installer.dsm7`:
```
#
# syncronize all files from var folder in spk (extracted to target/var)
# (@appdata) /var/packages/<package>/target/var -> /var/packages/<package>/var
#   
syno_sync_var_folder () {
    # take all files that do not alredy exist and rename the other files to *.new to avoid overwriting
    # existing configuration files.
    # finally remove the target/var folder that is not used under DSM 7.
    if [ -d ${SYNOPKG_PKGDEST}/var -a "$(ls -A ${SYNOPKG_PKGDEST}/var 2>/dev/null)" ]; then

        echo "Install files from var folder"

        # Move all files except existing
        echo "$RSYNC --ignore-existing --remove-source-files ${SYNOPKG_PKGDEST}/var/ ${SYNOPKG_PKGVAR}"
        $RSYNC --ignore-existing --remove-source-files ${SYNOPKG_PKGDEST}/var/ ${SYNOPKG_PKGVAR}

        # Rename any remaining (thus pre-existing) files with .new suffix
        find ${SYNOPKG_PKGDEST}/var -type f -exec sh -c 'x="{}"; mv "$x" "${x}.new"' \;
    
        # Move .new files (overwrite existing)
        echo "$RSYNC --remove-source-files ${SYNOPKG_PKGDEST}/var/ ${SYNOPKG_PKGVAR}"
        $RSYNC --remove-source-files ${SYNOPKG_PKGDEST}/var/ ${SYNOPKG_PKGVAR}

        # Remove var folder extraced from spk
        $RM ${SYNOPKG_PKGDEST}/var
    fi
}
```
 
> Not installing files to target/var is fixed for ejabberd is fixed in the subsequent commit.

Remaining issue from your change is that packages refererring to `rsc:var/<file>` in their `PLIST` simply won't be packaged anymore on DSM6.  Quick testing result in error - tested against `diyspk/lm-sensors`:
```
make[5]: Leaving directory '/home/spksrc/master/spksrc/diyspk/lm-sensors/work-aarch64-6.2.4/lm-sensors-3-6-0'
find /home/spksrc/master/spksrc/diyspk/lm-sensors/work-aarch64-6.2.4/install//var/packages/lm-sensors/target// \! -type d -printf '%P\n' | sed 's?^target/??g' | sort | \
  diff /home/spksrc/master/spksrc/diyspk/lm-sensors/work-aarch64-6.2.4/lm-sensors.plist.tmp -  | grep '>' | sed 's?> ??g' > /home/spksrc/master/spksrc/diyspk/lm-sensors/work-aarch64-6.2.4/lm-sensors.plist
make[4]: Leaving directory '/home/spksrc/master/spksrc/cross/lm-sensors'
make[3]: Leaving directory '/home/spksrc/master/spksrc/cross/lm-sensors'
===>  Creating target installation dir of lm-sensors
===>  [DSM6] Copy target to staging
tar: var/sensors3.conf: Cannot stat: No such file or directory
tar: Exiting with failure status due to previous errors
make[2]: *** [../../mk/spksrc.copy.mk:59: copy_target] Error 2
make[2]: Leaving directory '/home/spksrc/master/spksrc/diyspk/lm-sensors'
===>  20260111-150225 MAKELEVEL: 01, PARALLEL_MAKE: max, ARCH: aarch64-6.2.4, NAME: lm-sensors [END]
make[1]: *** [../../mk/spksrc.supported.mk:73: build-arch-aarch64-6.2.4] Error 1
make[1]: Leaving directory '/home/spksrc/master/spksrc/diyspk/lm-sensors'
make: *** [../../mk/spksrc.supported.mk:62: arch-aarch64-6.2.4] Error 2
```

The following fixes things and allows using the same behavior for all DSM:
```
spksrc@spksrc13:~/master/spksrc/mk$ git diff spksrc.copy.mk
diff --git a/mk/spksrc.copy.mk b/mk/spksrc.copy.mk
index 0a1c32b82..2181eec77 100644
--- a/mk/spksrc.copy.mk
+++ b/mk/spksrc.copy.mk
@@ -44,20 +44,15 @@ pre_copy_target: copy_msg
 copy_target: SHELL:=/bin/bash
 copy_target: .SHELLFLAGS := -o pipefail -c
 copy_target: $(PRE_COPY_TARGET) $(INSTALL_PLIST)
-ifeq ($(call version_ge, ${TCVERSION}, 7.0),1)
-	@$(MSG) [DSM7+] Copy target to staging, discard var directory
+	@$(MSG) "Copy target to staging, discard var directory"
 	@(mkdir -p $(STAGING_DIR) && cd $(STAGING_INSTALL_PREFIX) && tar cpf - $$(cat $(INSTALL_PLIST) | sed -e '/^.*:var\/.*/d' -e 's/^.*://g')) | \
 	  tar xpf - -C $(STAGING_DIR)
-	@$(MSG) "[DSM7+] Copy and merge var and target/var to $(STAGING_DIR)/var"
+	@$(MSG) "Copy and merge var and target/var to $(STAGING_DIR)/var"
 	@if [ "$$(cat $(INSTALL_PLIST) | sed -n 's?^.*:var/??p')" ] ; then \
 	  mkdir -p $(STAGING_DIR)/var ; \
 	  (cd $(STAGING_INSTALL_PREFIX)/../ && tar cpf - $$(eval ls -d $$(cat $(INSTALL_PLIST) | sed -n 's?^.*:var/??p' | sed -e 's?^?{var,target/var}/?') 2>/dev/null)) | \
 	    tar xpf - -C $(STAGING_DIR)/var --strip-components=1 --transform='s!^target/!!' ; \
 	fi
-else
-	@$(MSG) [DSM6] Copy target to staging
-	@(mkdir -p $(STAGING_DIR) && cd $(STAGING_INSTALL_PREFIX) && tar cpf - $$(cat $(INSTALL_PLIST) | cut -d':' -f2)) | tar xpf - -C $(STAGING_DIR)
-endif
 
 post_copy_target: $(COPY_TARGET)
 
```
I also suggest this `$(abspath $(INSTALL_PREFIX)/../var)` such as:
```
spksrc@spksrc13:~/master/spksrc/mk$ git diff spksrc.directories.mk
diff --git a/mk/spksrc.directories.mk b/mk/spksrc.directories.mk
index 6f9a39034..9e7bc5b9a 100644
--- a/mk/spksrc.directories.mk
+++ b/mk/spksrc.directories.mk
@@ -61,10 +61,8 @@ endif
 # when invoking make from under spk/*.  Setting var when
 # test-building dependencies from under cross/* is unecessary.
 #
-ifeq ($(call version_ge, ${TCVERSION}, 7.0),1)
 ifeq ($(lastword $(subst /, ,$(INSTALL_PREFIX))),target)
-INSTALL_PREFIX_VAR = $(shell dirname $(INSTALL_PREFIX))/var
-endif
+INSTALL_PREFIX_VAR = $(abspath $(INSTALL_PREFIX)/../var)
 endif
 ifeq ($(strip $(INSTALL_PREFIX_VAR)),)
 INSTALL_PREFIX_VAR  = $(INSTALL_PREFIX)/var
```

mreid-tt replied:
Superseded by #6906

mreid-tt committed:
Revert "fix INSTALL_PREFIX_VAR"

This reverts commit e0bcf9809adf9aa95c47106f6d17f986ab5ea949.

mreid-tt committed:
Merge branch 'master' into pr/5829

mreid-tt committed:
rabbitmq: Add CLI tools with proper permission handling

- Fix SPK_DEPENDS decimal format (erlang>=25.0:erlang<25.4)
- Add wrapper scripts for CLI tools (rabbitmqctl, rabbitmq-diagnostics,
  rabbitmq-plugins, rabbitmq-queues, rabbitmq-streams, rabbitmq-upgrade)
- CLI wrappers run as package user (sc-rabbitmq) via sudo to access
  the Erlang cookie (which must be mode 400 per Erlang requirements)
- Set HOME to SYNOPKG_PKGVAR in service-setup.sh and rabbitmq-env.conf
- Add SPK_COMMANDS for CLI tool symlinks in PATH
- Display user-friendly message when sudo password is required

mreid-tt committed:
Merge branch 'master' into pr/5829

mreid-tt commented:
hey @hgy59 — hope you don’t mind, I was trying to help get this one over the line and ready for publishing. I saw your earlier concern about RabbitMQ not working on DSM 7, so I did some upgrade testing on my end.

During testing, I found that `SPK_DEPENDS` needed consistent decimal version formatting; otherwise, upgrades would fail. I’ve adjusted that accordingly. I also added wrappers for the existing CLI tools to make RabbitMQ easier to test, and set `HOME` to `SYNOPKG_PKGVAR` so Erlang can correctly locate the cookie file.

Based on my DSM 7.1 testing, everything appears to be working now.

mreid-tt commented:
hey @hgy59, I did some further investigation and noted that for ejabberd the upgrade process doesn't seem to work properly. Here are the two scenarios I noted:

For DSM 6: The default admin is migrated but not enabled as an admin because this section is missing in the `ejabberd.yml` file:
```
acl:
  admin:
    user:
      - "admin@localhost"
```

For DSM 7: The `ejabberd.yml` file is also missing the addition of the above section but the default admin account does not seem to be migrated at all. In the log file (even when I manually add the above section), I see this:
```
[warning] <0.547.0>@ejabberd_web_admin:process2/2:215 Access of <<"admin@localhost">> from <<"::ffff:192.168.10.159">> failed with error: <<"inexistent-account">>
```

So there seems to be a bit more work needed where upgrades are concerned. New installs seem to work with DSM 6 and 7.

mreid-tt committed:
ejabberd: Add config-based admin ACL migration for upgrades

During upgrades, check if the existing ejabberd.yml has a properly
configured admin ACL. If the admin ACL section is missing or has no
valid user entries (or still has the @@adminuser@@ placeholder),
the upgrade wizard prompts the user for their existing admin credentials.

The admin account is validated against the ejabberd database using
ejabberdctl check_account before the upgrade proceeds, ensuring the
credentials match an actual account in the system.

This approach detects the actual configuration state regardless of
which version the user is upgrading from.

mreid-tt committed:
rabbitmq: Add DSM 7 upgrade data migration and fix CLI wrapper permission check

mreid-tt committed:
rabbitmq: Add web interface shortcut

mreid-tt committed:
ejabberd: Improve install wizard user experience

- Split wizard into two logical steps: Administrator Account and Server Domain
- Add introductory descriptions explaining ejabberd and XMPP
- Improve field labels for consistency with upgrade wizard
- Remove insecure default password (admin)
- Add minimum password length validation (4 characters)
- Clarify username format and login instructions
- Quote variables in register command to support passwords with spaces

Password validation follows XMPP ResourcePrep (RFC 6122) - special characters
are allowed; only non-printable and control characters are prohibited.

mreid-tt committed:
ejabberd: Support custom domain while preserving localhost access
